File: cpp/medium/0075_sort_colors.cpp
// https://leetcode.com/problems/sort-colors/description/?envType=daily-question&envId=2024-08-10
class Solution {
public:
    void sortColors(vector<int>& nums) {
        // No Extra Space | Two Pointers
        int i = 0;
        int j = nums.size() - 1;
        int k = 0;
        for(int k = 0; k <= j; k++){
            if(nums[k] == 0){
                swap(nums[i], nums[k]);
                i++;
            }
            if(nums[k] == 2){
                swap(nums[j], nums[k]);
                k--;
                j--;
            }
        }

        // // Extra Space Solution
        // int count[3];
        // for(int num : nums){
        //     count[num]++;
        // }
        // int j = 0;
        // for(int i = 0; i < nums.size(); i++){
        //     if(count[j] == 0){
        //         j++;
        //         i--;
        //     } else{
        //         nums[i] = j;
        //         count[j]--;
        //     }
        // }
    }
};

File: python/medium/0075_sort_colors.py
# https://leetcode.com/problems/sort-colors/description/?envType=daily-question&envId=2025-05-17
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        start = 0
        end = len(nums) - 1
        i = 0
        while i <= end:
            if nums[i] == 0:
                if i != start:
                    nums[i], nums[start] = nums[start], nums[i]
                    start += 1
                    i -= 1
            elif nums[i] == 2:
                if i != end:
                    nums[i], nums[end] = nums[end], nums[i]
                    end -= 1
                    i -= 1
            i += 1
        return nums
